### PR TITLE
Clean up determination of config file default path

### PR DESCRIPTION
**Pull request by mattias-p**
## Purpose

This is mainly a code quality improvement. It also avoids locating the default backend_config.ini if it isn't needed.

## Context

N/A

## Changes

* It avoids executing some code at load time.
* It replaces a global variable (`$Zonemaster::Backend::Config::path`) with a procedure (`Zonemaster::Backend::Config::get_default_path()`).
* It adds code documentation about where the default backend_config.ini file is sought.
* It simplifies the code to determine the config file path.

## How to test this PR

This should not change the behavior of Backend in production context.

A unit test has been added to make sure the Zonemaster::Backend::Config module can be loaded without it being able to locate a backend_config.ini file at load time.

### CODE REVIEW

**Review by pnax — APPROVED** [review]

**Comment by matsduf** [comment]
Is there really any good reason to have a default path the configuration file? Backend is always started by start script and in that the environment variable can be set, unless already set. By not having any default file the risk that backend is started by an unintended configuration file lowers and it is also less complex.

**Comment by mattias-p** [comment]
It might make sense to do something in that direction. But the situation is more complex than just starting the daemons from the service scripts. There's also the database creation and migration scripts. And the unit tests. And maybe something else I'm forgetting. I suggest you create a new issue for it.

**Comment by matsduf** [comment]
> It might make sense to do something in that direction. But the situation is more complex than just starting the daemons from the service scripts. There's also the database creation and migration scripts. And the unit tests. And maybe something else I'm forgetting. I suggest you create a new issue for it.

I think database creation and migration script underlines the problem with the built in default configuration files, escpecially "DIST_DIR/backend_config.ini" since it is not really supposed to be used by an installed Backend.

**Comment by pnax** [comment]
The new test file needs to be added to the Manifest

**Comment by mattias-p** [comment]
> I think database creation and migration script underlines the problem with the built in default configuration files, escpecially "DIST_DIR/backend_config.ini" since it is not really supposed to be used by an installed Backend.

Agreed. But that's out of scope for this PR.

I added the unit test to the manifest and reran Travis.

Please re-review.

**Review by pnax — APPROVED** [review]

File: lib/Zonemaster/Backend/Config.pm [hunk 1]
```diff
@@ -15,27 +15,52 @@ use Readonly;
...
-}
+=head2 get_default_path
 
-Readonly my @SIG_NAME => split ' ', $Config{sig_name};
+Determine the path for the default backend_config.ini file.
+A list of values and locations are checked and the first match is returned.
+If all places are checked and no file is found, an exception is thrown.
+
+This procedure is idempotent - i.e. if you call this procedure multiple times
+the same value is returned no matter if envirnmental circumstances have changed.
```
Comment (matsduf) [inline lib/Zonemaster/Backend/Config.pm]: I do not understand this. Please explain.

Say that the environment variable is unset, but /etc/zonemaster/backend_config.ini exists, then get_default_path() would return "/etc/zonemaster/backend_config.ini".

If then "/etc/zonemaster/backend_config.ini" is removed, would get_default_path() still return "/etc/zonemaster/backend_config.ini".

Please re-write in simplified language.

**Review by matsduf — CHANGES_REQUESTED** [review]

File: lib/Zonemaster/Backend/Config.pm (hunk 1 above)
Comment (mattias-p) [reply lib/Zonemaster/Backend/Config.pm]: Is it better now?

File: lib/Zonemaster/Backend/Config.pm [hunk 2]
```diff
@@ -15,27 +15,53 @@ use Readonly;
...
+
+If a file exists at this path, it is returned.
+DIST_DIR is wherever File::ShareDir installs the Zonemaster-Backend dist.
+
+=back
+
+=cut
+
+sub get_default_path {
+    state $path =
```
Comment (matsduf) [inline lib/Zonemaster/Backend/Config.pm:58]: I cannot see that `use feature state` is set. Isn't that necessary?

File: lib/Zonemaster/Backend/Config.pm (hunk 1 above)
Comment (matsduf) [reply lib/Zonemaster/Backend/Config.pm]: Yes, thanks.

File: lib/Zonemaster/Backend/Config.pm (hunk 2 above)
Comment (mattias-p) [reply lib/Zonemaster/Backend/Config.pm:58]: It's enabled by default since Perl 5.10 and we're declaring a minimum version of 5.14.

**Review by matsduf — APPROVED** [review]

**Review by blacksponge — APPROVED** [review]

**Pull request merged**